File: src/linked lists/Fundamentals Single LL/insert_an_elt_before_x.cpp
#include <bits/stdc++.h>
using namespace std;

struct Node {
    int data;
    Node* next;

    Node(int data) {
        this->data = data;
        this->next = nullptr;
    }

    Node(int data, Node* next) {
        this->data = data;
        this->next = next;
    }
};

class Solution {
   public:
    Node* insertBeforeX(Node* head, int el, int x) {
        if (head == nullptr) {
            return head;
        }

        if (head->data == x) {
            Node* newNode = new Node(el, head);
            return newNode;
        }

        Node* prev = nullptr;
        Node* temp = head;
        while (temp != NULL) {
            if (temp->data == x) {
                Node* newNode = new Node(el, temp);
                prev->next = newNode;
                break;
            }
            prev = temp;
            temp = temp->next;
        }
        return head;
    }
};

int main() { return 0; }

File: src/linked lists/Fundamentals Single LL/insert_an_elt_before_x.js
class Node {
    constructor(x, next = null) {
        this.data = x;
        this.next = null;
    }
};

class Solution {
    insertBeforeX(head, el, x) {
        if (head == null) return null;

        let prev = null;
        let temp = head;

        // can't access next of null so have to handle this case separately
        if (head.data === el) {
            let newNode = new Node(el);
            newNode.next = head;
            return newNode;
        }

        while (temp != null) {
            if (temp.data == x) {
                let newNode = new Node(el);
                prev.next = newNode;
                newNode.next = temp;
                break;
            }
            prev = temp;
            temp = temp.next;
        }
        return head;
    }
};

function printLL(head) {
    let temp = head;
    while (temp !== null) {
        console.log(temp.data + " ");
        temp = temp.next;
    }
}

function main() {
    // make a LL
    let head = new Node(3);
    head.next = new Node(1);
    head.next.next = new Node(5);
    head.next.next.next = new Node(8);

    let sol = new Solution();
    head = sol.insertBeforeX(head, 0, 5);

    printLL(head);
}

main();